src/object/srv_ec_aggregate.c

```diff
@@ -1468,85 +1626,102 @@ agg_process_holes_ult(void *arg)
+						peer);
+					rc = -1;
+					goto out;
+				}
 			}
 		}
-	}
 
-	tgt_ep.ep_rank = entry->ae_peer_pshards[0].sd_rank;
-	tgt_ep.ep_tag = entry->ae_peer_pshards[0].sd_tgt_idx;
-	rc = obj_req_create(dss_get_module_info()->dmi_ctx, &tgt_ep,
-			    DAOS_OBJ_RPC_EC_REPLICATE, &rpc);
-	if (rc) {
-		D_ERROR("obj_req_create failed: "DF_RC"\n", DP_RC(rc));
-		goto out;
-	}
-	ec_rep_in = crt_req_get(rpc);
-	agg_param = container_of(entry, struct ec_agg_param, ap_agg_entry);
-	uuid_copy(ec_rep_in->er_pool_uuid,
-		  agg_param->ap_pool_info.api_pool_uuid);
-	uuid_copy(ec_rep_in->er_coh_uuid,
-		  agg_param->ap_pool_info.api_poh_uuid);
-	uuid_copy(ec_rep_in->er_cont_uuid,
-		  agg_param->ap_pool_info.api_cont_uuid);
-	uuid_copy(ec_rep_in->er_coh_uuid,
-		  agg_param->ap_pool_info.api_coh_uuid);
-	ec_rep_in->er_oid = entry->ae_oid;
-	ec_rep_in->er_oid.id_shard--;
-	ec_rep_in->er_dkey = entry->ae_dkey;
-	ec_rep_in->er_iod = iod;
-	ec_rep_in->er_stripenum = entry->ae_cur_stripe.as_stripenum;
-	ec_rep_in->er_epoch = entry->ae_cur_stripe.as_hi_epoch;
-	ec_rep_in->er_map_ver =
-		agg_param->ap_pool_info.api_pool->sp_map_version;
-	entry->ae_sgl.sg_nr_out = 1;
-	rc = crt_bulk_create(dss_get_module_info()->dmi_ctx, &entry->ae_sgl,
-			     CRT_BULK_RW, &ec_rep_in->er_bulk);
-	if (rc) {
-		D_ERROR("crt_bulk_create returned: "DF_RC"\n", DP_RC(rc));
-		goto out;
-	}
-	rc = dss_rpc_send(rpc);
-	if (rc) {
-		D_ERROR("dss_rpc_send failed: "DF_RC"\n", DP_RC(rc));
-		goto out;
+		tgt_ep.ep_rank = entry->ae_peer_pshards[peer].sd_rank;
+		tgt_ep.ep_tag = entry->ae_peer_pshards[peer].sd_tgt_idx;
+		rc = obj_req_create(dss_get_module_info()->dmi_ctx, &tgt_ep,
+				    DAOS_OBJ_RPC_EC_REPLICATE, &rpc);
+		if (rc) {
+			D_ERROR(DF_UOID" obj_req_create failed: "DF_RC"\n",
+				DP_UOID(entry->ae_oid), DP_RC(rc));
+			goto out;
+		}
+		ec_rep_in = crt_req_get(rpc);
+		uuid_copy(ec_rep_in->er_pool_uuid,
+			  agg_param->ap_pool_info.api_pool_uuid);
+		uuid_copy(ec_rep_in->er_coh_uuid,
+			  agg_param->ap_pool_info.api_poh_uuid);
+		uuid_copy(ec_rep_in->er_cont_uuid,
+			  agg_param->ap_pool_info.api_cont_uuid);
+		uuid_copy(ec_rep_in->er_coh_uuid,
+			  agg_param->ap_pool_info.api_coh_uuid);
+		ec_rep_in->er_oid = entry->ae_oid;
+		ec_rep_in->er_oid.id_shard--;
+		ec_rep_in->er_dkey = entry->ae_dkey;
+		ec_rep_in->er_iod = iod;
+		ec_rep_in->er_stripenum = entry->ae_cur_stripe.as_stripenum;
+		ec_rep_in->er_epoch = entry->ae_cur_stripe.as_hi_epoch;
+		ec_rep_in->er_map_ver =
+			agg_param->ap_pool_info.api_pool->sp_map_version;
+		ec_rep_in->er_bulk = bulk_hdl;
+		rc = dss_rpc_send(rpc);
+		if (rc) {
+			D_ERROR(DF_UOID" peer %d dss_rpc_send failed "DF_RC"\n",
+				DP_UOID(entry->ae_oid), peer, DP_RC(rc));
+			goto out;
+		}
+		ec_rep_out = crt_reply_get(rpc);
+		rc = ec_rep_out->er_status;
+		crt_req_decref(rpc);
+		rpc = NULL;
+		if (rc) {
+			D_ERROR(DF_UOID" peer %d er_status failed "DF_RC"\n",
+				DP_UOID(entry->ae_oid), peer, DP_RC(rc));
+			break;
```

Comment:
this seems duplicate with peer_update_ult() ?

Reply:
right, we may consider to combine those two and reuse most of the code, in following patch.
This patch just here just add support for P > 2 case, and I tested 4+3 seems work fine.

Reply:
Agree with this for future change.  My version ended up duplicating a lot of code for the two RPCs.